compiler/rustc_ty_utils/src/opaque_types.rs

```diff
@@ -35,33 +52,91 @@ impl<'tcx> OpaqueTypeCollector<'tcx> {
         })
     }
 
-    fn visit_spanned(&mut self, span: Span, value: impl TypeVisitable<TyCtxt<'tcx>>) {
+    fn visit_spanned_after_normalizing(
+        &mut self,
+        span: Span,
+        value: impl TypeFoldable<TyCtxt<'tcx>>,
+    ) {
         let old = self.span;
         self.span = Some(span);
-        value.visit_with(self);
-        self.span = old;
-    }
 
-    fn parent_trait_ref(&self) -> Option<ty::TraitRef<'tcx>> {
-        let parent = self.parent()?;
-        if matches!(self.tcx.def_kind(parent), DefKind::Impl { .. }) {
-            Some(self.tcx.impl_trait_ref(parent)?.instantiate_identity())
-        } else {
-            None
+        if let Ok(value) = self.normalize_if_possible(span, value) {
+            value.visit_with(self);
         }
```

Comment:
hm... so if we fail to normalize, we don't visit at all...

Please add a test that shows that you can use projections to mention a TAIT in a signature without it showing up in the signature and thus being defining. I believe we are already testing the case where failing to normalize ignores all TAITs, even those in the projection.

Reply:
Hmmm... what do you mean by this exactly?

Reply:
I was thinking about a `<Tait as Trait>::Assoc` projection that is know to result in `()` or something that does not contain the `Tait`, but I think we already have such tests, as I added some a while ago